fragmenter/chemi.py

```diff
@@ -1,1619 +1,525 @@
+            ):
+                continue
 
+            new_valence -= neighbour_bond.GetValenceContrib(atom)
 
-def mol_to_image_atoms_label(mol, fname, map_idx=True, width=600, height=400, label_scale=2.0, scale_bondwidth=True):
+        if numpy.isclose(old_valence, new_valence):
+            # Skip the cases where the valence won't change
+            continue
 
-    """Write out png file of molecule with atoms labeled with their map index.
+        if (
+            atom.GetAtomicNum() == 6
+            and atom.GetIsAromatic()
+            and sum(
+                1 for bond_tuple in bond_indices if atom.GetAtomMapNum() in bond_tuple
+            )
+            == 1
+        ):
 
-    Parameters
-    ----------
-    smiles: str
-        SMILES
-    fname: str
-        absolute path and filename for png
-    map_idx: bool
-        If True, lable atoms with map index instead of atom index. If set to True, input SMILES must have map indices.
+            # This is likely a cap carbon which was retained from an existing ring. It's
+            # aromaticity needs to be cleared before calling ``MolFragmentToSmiles``
+            # otherwise will (understandably) be confused and throw an exception.
+            atom.SetIsAromatic(False)
 
-    """
-    if not isinstance(mol, oechem.OEMol) and isinstance(mol, str):
-        # Try converting smiles to mol
-        mol = cmiles.utils.load_molecule(mol, toolkit='openeye')
+        # Add a hydrogen to the atom whose valence will change.
+        for _ in range(int(numpy.rint(old_valence - new_valence))):
 
-    mol = oechem.OEMol(mol)
-    oedepict.OEPrepareDepiction(mol)
-    opts = oedepict.OE2DMolDisplayOptions(width, height, oedepict.OEScale_AutoScale)
+            new_atom = Chem.Atom(1)
+            new_atom_index = rd_molecule.AddAtom(new_atom)
 
-    if map_idx:
-        # check if molecule has map
-        if not cmiles.utils.has_atom_map(mol):
-            raise ValueError("Input SMILES must have atom maps to display map indices in image")
-        opts.SetAtomPropertyFunctor(oedepict.OEDisplayAtomMapIdx())
-        opts.SetAtomPropertyFunctor(oedepict.OEDisplayAtomMapIdx())
-    if not map_idx:
-        opts.SetAtomPropertyFunctor(oedepict.OEDisplayAtomIdx())
+            rd_molecule.AddBond(atom_index, new_atom_index)
 
-    opts.SetAtomPropLabelFont(oedepict.OEFont(oechem.OEDarkGreen))
-    opts.SetAtomPropLabelFontScale(label_scale)
-    opts.SetBondWidthScaling(scale_bondwidth)
+            new_bond = rd_molecule.GetBondBetweenAtoms(atom_index, new_atom_index)
+            new_bond.SetBondType(Chem.BondType.SINGLE)
+            new_bond.SetIsAromatic(False)
 
-    disp = oedepict.OE2DMolDisplay(mol, opts)
-    return oedepict.OERenderMolecule(fname, disp)
+            atoms_to_use.append(new_atom_index)
+            bonds_to_use.append(new_bond.GetIdx())
 
+    fragment_smiles = Chem.MolFragmentToSmiles(rd_molecule, atoms_to_use, bonds_to_use)
```

Comment:
Thanks for introducing me to this method, this seems really useful. WRT atom maps I see by default in the [docs](https://www.rdkit.org/docs/source/rdkit.Chem.rdmolfiles.html#rdkit.Chem.rdmolfiles.MolFragmentToSmiles) this method will attempt to put the molecule in canonical order, I was thinking this could affect the atom map should this be set to false? As far as the openeye equivalent goes it does not mention changing the ordering which made me think this might not be what we want.

Reply:
> I was thinking this could affect the atom map should this be set to false? As far as the openeye equivalent goes it does not mention changing the ordering which made me think this might not be what we want.

Hmm thanks for pointing that out! I think because the map is explicitly included in the produced SMILES pattern it shouldn't be affected by a canonical re-ordering, although there doesn't seem to be a downside to disabling the canonical ordering so I will try this out.